packs/BP/scripts/network.ts

```diff
@@ -89,123 +93,152 @@ export class MachineNetwork extends DestroyableObject {
-            | InternalRegisteredMachine
-            | undefined;
+    // initialize consumers keys.
+    const consumers: Record<string, ConsumerGroups> = {};
+    for (const key of typesToDistribute) {
+      consumers[key] = { lowPriority: [], normalPriority: [] };
+    }
 
-          if (!definition) {
-            logWarn(
-              `can't send '${queuedSend.type}' to machine '${block.typeId}': this block is configured as a machine but it couldn't be found in the machine registry`,
-            );
-            yield;
-            continue;
-          }
+    // find and filter connections into their consumer groups.
+    this.connections.machines.forEach((machine) => {
+      const machineTags = machine.getTags();
+      const isLowPriority = machineTags.includes(
+        "fluffyalien_energisticscore:low_priority_consumer",
+      );
+      const allowsAny = machineTags.includes(
+        "fluffyalien_energisticscore:consumer._any",
+      );
 
-          const amount = getMachineStorage(block, queuedSend.type);
+      // Check machine tags and sort into appropriate groups.
+      Object.keys(consumers).forEach((consumerType) => {
+        const allowsType =
+          allowsAny ||
+          machineTags.includes(
+            `fluffyalien_energisticscore:consumer.${consumerType}`,
+          );
+        if (!allowsType) return;
+
+        if (isLowPriority) consumers[consumerType].lowPriority.push(machine);
+        else consumers[consumerType].normalPriority.push(machine);
+      });
+    });
+
+    // send each machine its share of the pool.
+    for (const type of typesToDistribute) {
+      const machines = consumers[type];
+      const numMachines =
+        machines.lowPriority.length + machines.normalPriority.length;
+      if (numMachines === 0) continue;
+
+      let budget = distribution[type];
+
+      // Give each machine in the normal priority an equal split of the budget
+      // Machines can consume less than they're offered in which case the savings are given to further machines.
+      for (let i = 0; i < machines.normalPriority.length; i++) {
+        const machine = machines.normalPriority[i];
+        const budgetAllocation = budget / (machines.normalPriority.length - i);
+        const currentStored = getMachineStorage(machine, type);
+        const machineDef = getMachineRegistration(machine);
+
+        let amountToAllocate: number = Math.min(
+          budgetAllocation,
+          machineDef.maxStorage - currentStored,
+        );
 
-          if (amount >= definition.maxStorage) {
-            yield;
-            continue;
-          }
+        let waiting = true;
 
-          targetsArr.push({
-            block,
-            amount,
-            definition,
+        this.sendMachineAllocation(machine, machineDef, type, amountToAllocate)
+          .then((v) => {
+            amountToAllocate = v;
+            waiting = false;
+          })
+          .catch((e: unknown) => {
+            throw e;
```

Comment:
If there is no catch handler then the promise will throw, no need to catch the error then rethrow. Remove `.catch` and `void` the promise to mark it as ignored.
```ts
void this.sendMachineAllocation(machine, machineDef, type, amountToAllocate)
    .then((v) => {
        amountToAllocate = v;
        waiting = false;
    });
```

Reply:
okay one other thing i just realized, the generator will never end if an error is thrown from the promise, even if it's from the `.catch`. it has to be thrown from directly from the generator function.

so you have to do something like this
```ts
let shouldThrow = false;
this.sendMachineAllocation(machine, machineDef, type, amountToAllocate)
    .then((v) => {
        amountToAllocate = v;
        waiting = false;
    })
    .catch((e) => {
      waiting = false;
      shouldThrow = true;
    })
while (waiting) yield;
if (shouldThrow) {
  throw new Error(...)
}
```
**or just log a warning and set the allocation to 0**